client/mixnet/src/request.rs

```diff
@@ -0,0 +1,112 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Sender-side request logic. Some things from this module are also used on the receiver side, eg
+//! [`extrinsic_delay`], but most of the receiver-side request logic lives elsewhere.
+
+use super::{config::SubstrateConfig, error::Error};
+use blake2::{
+	digest::{consts::U16, Mac},
+	Blake2bMac,
+};
+use codec::{Decode, DecodeAll};
+use futures::channel::oneshot;
+use log::warn;
+use mixnet::core::{Delay, MessageId, PostErr, Scattered};
+use sp_core::Bytes;
+use std::time::Duration;
+
+const LOG_TARGET: &str = "mixnet";
+
+fn send_err<T>(reply_sender: oneshot::Sender<Result<T, Error>>, err: Error) {
+	if let Err(Err(err)) = reply_sender.send(Err(err)) {
+		warn!(target: LOG_TARGET, "Failed to inform requester of error: {err}");
+	}
+}
+
+fn send_reply<T: Decode>(reply_sender: oneshot::Sender<Result<T, Error>>, mut data: &[u8]) {
+	let res = match Result::decode_all(&mut data) {
+		Ok(res) => res.map_err(Error::Remote),
+		Err(_) => Err(Error::BadReply),
+	};
+	match reply_sender.send(res) {
+		Ok(_) => (),
+		Err(Ok(_)) => warn!(target: LOG_TARGET, "Failed to send reply to requester"),
+		Err(Err(err)) => warn!(target: LOG_TARGET, "Failed to inform requester of error: {err}"),
+	}
+}
+
+pub const SUBMIT_EXTRINSIC: u8 = 1;
```

Comment:
Mainly because it's simpler and more efficient to just handle the top-level enum manually. Note that the encoding _is_ effectively SCALE, the library just isn't used for handling the top-level enum.

Reply:
I'm not really sure we need this small benefit of being more performant vs better readability.

Reply:
I don't agree about the readability honestly. It's straight forward as-is: the client side sticks a `SUBMIT_EXTRINSIC` byte at the start and the server side does a match on the first byte. In my eyes that is simpler and easier to understand than encode/decode calls.

Note that with the `Request` enum as you have defined it you are going to effectively double-encode the extrinsic, which is a bit confusing IMO.